### PR TITLE
Add direct tests for predicate pushdown optimizer

### PR DESCRIPTION
**Pull request by tlento**
This addresses the testing gap around the cases covered by the
logic internal to the PredicatePushdownOptimizer.

It's rather difficult to hand-construct dataflow plans, so these
rely on snapshots of DataflowPlan text output for some carefully
selected queries which are known to cover the code paths in question.

### CODE REVIEW

**Comment by tlento** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1303?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#1303** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1303?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#1302** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1302?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1301** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1301?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1300** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1300?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1286** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1286?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1280** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1280?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1279** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1279?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1278** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1278?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1277** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1277?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1263** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1263?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1262** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1262?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1240** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1240?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1227** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1227?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1230** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1226** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1226?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1225** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1225?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1224** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1224?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1218** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1218?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1216** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1216?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#1217](https://github.com/dbt-labs/metricflow/pull/1217) <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1217?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#1215** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1215?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1214** <a href="https://app.graphite.dev/github/pr/dbt-labs/metricflow/1214?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @tlento and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by courtneyholcomb — APPROVED** [review]
LGTM!

**Comment by courtneyholcomb** [comment]
My thought on more edge cases to test: you could test each of the ones you already added, but nested in a derived metric.

File: tests_metricflow/dataflow/optimizer/test_predicate_pushdown_optimizer.py [hunk 1]
```diff
@@ -148,3 +160,238 @@ def test_applied_filter_back_propagation(branch_state_tracker: PredicatePushdown
...
+    _check_optimization(
+        request=request,
+        mf_test_configuration=mf_test_configuration,
+        dataflow_plan_builder=dataflow_plan_builder,
+        query_spec=query_spec,
+        expected_additional_constraint_nodes_in_optimized=0,
+    )
+
+
+def test_offset_metric_predicate_pushdown(
```
Comment (courtneyholcomb) [inline tests_metricflow/dataflow/optimizer/test_predicate_pushdown_optimizer.py:315]: I think we also have an extra where filter applied to cumulative metrics with `grain_to_date`, so might be good to test that specifically!

Comment (tlento) [reply tests_metricflow/dataflow/optimizer/test_predicate_pushdown_optimizer.py:315]: As far as I can tell those are only for time constraints, not general where constraints, but I'll add a test case to the top of the stack.

**Comment by tlento** [comment]
> My thought on more edge cases to test: you could test each of the ones you already added, but nested in a derived metric.

I'm relying on our existing derived metric coverage (the nested filter test has been instructive), but yes, this is a good idea. I'll think about how to expand scope - setting up the filter sequencing gets pretty nasty.

**Comment by tlento** [comment]
### Merge activity

* **Jun 25, 5:24 PM PDT**: @tlento started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dbt-labs/metricflow/1303).
